src/v/io/tests/cache_test.cc

```diff
@@ -0,0 +1,728 @@
+        testing_details::cache_hook_accessor::set_hook_insertion_time(
+          entry.hook, time);
+    }
+
+    static uint8_t get_hook_freq(entry& entry) {
+        return testing_details::cache_hook_accessor::get_hook_freq(entry.hook);
+    }
+
+    template<typename... Entries>
+    void cleanup(Entries&... entries) {
+        (cache->remove(entries), ...);
+    }
+
+    std::unique_ptr<cache_type> cache;
+};
+
+TEST_F(CacheTest, DefaultState) {
+    EXPECT_EQ(cache->stat().main_queue_size, 0);
+    EXPECT_EQ(cache->stat().small_queue_size, 0);
+}
+
+TEST_F(CacheTest, DefaultHookState) {
+    entry e;
+    EXPECT_TRUE(e.hook.evicted());
+    EXPECT_FALSE(cache->on_ghost_queue(e));
+
+    cleanup(e);
+}
+
+TEST_F(CacheTest, HookCopy) {
+    entry e0;
+    e0.hook.touch();
+    e0.hook.touch();
+    cache->insert(e0);
+    set_hook_insertion_time(e0, 3); // after insert since insert will reset
+    EXPECT_EQ(get_hook_freq(e0), 2);
+    EXPECT_EQ(get_hook_insertion_time(e0), 3);
+    EXPECT_FALSE(e0.hook.evicted());
+
+    entry e1{.hook = e0.hook};
+    EXPECT_EQ(get_hook_freq(e1), 2);
+    EXPECT_EQ(get_hook_insertion_time(e1), 3);
+    EXPECT_TRUE(e1.hook.evicted());
+
+    cleanup(e0, e1);
+}
+
+TEST_F(CacheTest, Insert) {
+    entry e0;
+    cache_insert_main(e0);
+    EXPECT_EQ(cache->stat().main_queue_size, 1);
+    EXPECT_EQ(cache->stat().small_queue_size, 0);
+    EXPECT_FALSE(cache->on_ghost_queue(e0));
+    EXPECT_FALSE(e0.hook.evicted());
+
+    entry e1;
+    cache_insert_small(e1);
+    EXPECT_EQ(cache->stat().small_queue_size, 1);
+    EXPECT_EQ(cache->stat().main_queue_size, 1);
+    EXPECT_FALSE(cache->on_ghost_queue(e1));
+    EXPECT_FALSE(e1.hook.evicted());
+
+    cleanup(e0, e1);
+}
+
+TEST_F(CacheTest, InsertPastCapacity) {
+    // fill main queue to capacity
+    std::vector<entry> mains(main_size);
+    for (auto& e : mains) {
+        cache_insert_main(e);
+    }
+
+    // fill small queue to capacity
+    std::vector<entry> smalls(small_size);
+    for (auto& e : smalls) {
+        cache->insert(e);
+    }
+
+    EXPECT_EQ(
+      cache->stat().small_queue_size + cache->stat().main_queue_size,
+      cache_size);
+
+    // the next insertion will exceed capacity, and then
+    // subsequent insertions will first evict before insert.
+    std::vector<entry> more(4);
+    for (auto& e : more) {
+        cache->insert(e);
+        EXPECT_EQ(
+          cache->stat().small_queue_size + cache->stat().main_queue_size,
+          cache_size + 1);
+    }
+
+    // mark everything as non-evictable
+    for (auto& e : mains) {
+        e.may_evict = false;
+    }
+    for (auto& e : smalls) {
+        e.may_evict = false;
+    }
+    for (auto& e : more) {
+        e.may_evict = false;
+    }
+
+    // if everythign is non-evictable we'll allow insertion to continue. the
+    // assumption here is that this is one extremely rare and two that the
+    // caller has already made a memory reservation / accounted for the entry.
```

Comment:
Ah I see this is the answer to my previous questions... I'll mark those as resolved.